src/components/settings/PushNotifications/logic.ts

```diff
@@ -0,0 +1,123 @@
+import { keccak256, toUtf8Bytes } from 'ethers/lib/utils'
+import { getToken, getMessaging } from 'firebase/messaging'
+import { DeviceType } from '@safe-global/safe-gateway-typescript-sdk'
+import type { RegisterNotificationsRequest } from '@safe-global/safe-gateway-typescript-sdk'
+import type { Web3Provider } from '@ethersproject/providers'
+
+import { FIREBASE_VAPID_KEY, initializeFirebaseApp } from '@/services/push-notifications/firebase'
+import { trackEvent } from '@/services/analytics'
+import { PUSH_NOTIFICATION_EVENTS } from '@/services/analytics/events/push-notifications'
+import packageJson from '../../../../package.json'
+import { logError } from '@/services/exceptions'
+import ErrorCodes from '@/services/exceptions/ErrorCodes'
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
+
+// We store UUID locally to track device registration
+export type NotificationRegistration = WithRequired<RegisterNotificationsRequest, 'uuid'>
+
+export const requestNotificationPermission = async (): Promise<boolean> => {
+  if (Notification.permission === 'granted') {
+    return true
+  }
+
+  let permission: NotificationPermission | undefined
+
+  try {
+    permission = await Notification.requestPermission()
+  } catch (e) {
+    logError(ErrorCodes._400, e)
+  }
+
+  const isGranted = permission === 'granted'
+
+  trackEvent(isGranted ? PUSH_NOTIFICATION_EVENTS.GRANT_PERMISSION : PUSH_NOTIFICATION_EVENTS.REJECT_PERMISSION)
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
+
+  return isGranted
+}
+
+const getSafeRegistrationSignature = ({
+  safeAddresses,
+  web3,
+  timestamp,
+  uuid,
+  token,
+}: {
+  safeAddresses: Array<string>
+  web3: Web3Provider
+  timestamp: string
+  uuid: string
+  token: string
+}) => {
+  const MESSAGE_PREFIX = 'gnosis-safe'
+
+  const message = MESSAGE_PREFIX + timestamp + uuid + token + safeAddresses.join('')
+  const hashedMessage = keccak256(toUtf8Bytes(message))
+
+  return web3.getSigner().signMessage(hashedMessage)
+}
+
+export type NotifiableSafes = { [chainId: string]: Array<string> }
+
+export const getRegisterDevicePayload = async ({
+  safesToRegister,
+  uuid,
+  web3,
+}: {
+  safesToRegister: NotifiableSafes
+  uuid: string
+  web3: Web3Provider
+}): Promise<RegisterNotificationsRequest> => {
+  const BUILD_NUMBER = '0' // Required value, but does not exist on web
+  const BUNDLE = 'safe'
+
+  const [serviceWorkerRegistration] = await navigator.serviceWorker.getRegistrations()
+
+  // Get Firebase token
+  const app = initializeFirebaseApp()
+  const messaging = getMessaging(app)
+
+  const token = await getToken(messaging, {
```

Comment:
The documentation of this function says: `If a notification permission isn't already granted, this method asks the user for permission.`

Isn't that what we also do manually in `requestNotificationPermission`?

Reply:
Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

Reply:
See above comment.